pkg/scheduler/queue/tree_queueing_algorithms.go

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+// DequeueAlgorithm represents the set of operations specific to different approaches to dequeuing. It is applied
+// at the layer-level -- every Node at the same depth in a TreeQueue shares the same DequeueAlgorithm, including state
+// that may be stored in a struct that implements DequeueAlgorithm.
+type DequeueAlgorithm interface {
+	addChildNode(parent, child *Node)
+	dequeueSelectNode(node *Node) (*Node, bool)
+	dequeueUpdateState(node *Node, dequeuedFrom *Node)
+}
+
+// roundRobinState is the simplest type of DequeueAlgorithm; nodes which use this DequeueAlgorithm and are at
+// the same depth in a TreeQueue do not share any state. When children are added to these nodes, they are placed at
+// the "end" of the order from the perspective of the node's current queuePosition (e.g., if queuePosition is 3,
+// a new child will be placed at index 2). Children are dequeued from using a simple round-robin ordering;
+// queuePosition is incremented on every dequeue.
+type roundRobinState struct {
+}
+
+// addChildNode creates a new child Node and adds it to the "end" of the parent's queueOrder from the perspective
+// of the parent's queuePosition. Thus, If queuePosition is localQueueIndex, the new child is added to the end of
+// queueOrder. Otherwise, the new child is added at queuePosition - 1, and queuePosition is incremented to keep
+// it pointed to the same child.
+func (rrs *roundRobinState) addChildNode(parent, child *Node) {
+	// add childNode to n.queueMap
+	parent.queueMap[child.Name()] = child
```

Comment:
This feels out of place here, and I notice it exists in both `roundRobinState` and `tenantQuerierAssignments`. Why should the "Dequeue Algorithm" be responsible for adding a child node to the tree?

Reply:
I think the naming got changed a lot and just needs to be synced up, it's more like "Queue Algorithm" because adding the child node only happens during enqueue

Reply:
Yup, I've gone back and forth on naming a number of times while trying to keep changes tightly scoped, happy to rename the interface